src/core/Akka/IO/TcpConnection.cs

```diff
@@ -435,692 +258,523 @@ protected void CompleteConnect(IActorRef commander, IEnumerable<Inet.SocketOptio
-        {
-            while (_pendingAcks.TryDequeue(out var ackInfo))
+            if(_handler != null)
             {
-                ackInfo.Commander.Tell(ackInfo.Ack);
+                closeInformation = closeInformation with { NotificationsTo = closeInformation.NotificationsTo.Add(_handler!) };
             }
-
-            ClearStatus(ConnectionStatus.Sending);
+            
+            _closeInformation = closeInformation;
+            Context.Stop(Self);
         }
-
-        private void DoRead(ConnectionInfo info, IActorRef closeCommander)
-        {
-            //TODO: What should we do if reading is suspended with an oustanding read - this will discard the read
-            //      Should probably have an 'oustanding read'
-            if (!HasStatus(ConnectionStatus.ReadingSuspended))
-            {
-                try
-                {
-                    var read = InnerRead(info, Tcp.Settings.ReceivedMessageSizeLimit, ReceiveArgs);
-                    ClearStatus(ConnectionStatus.Receiving);
-                    switch (read.Type)
-                    {
-                        case ReadResultType.AllRead:
-                            if (!_pullMode)
-                                ReceiveAsync();
-                            break;
-                        case ReadResultType.EndOfStream:
-                            if (_isOutputShutdown)
-                            {
-                                if (_traceLogging) Log.Debug("Read returned end-of-stream, our side already closed");
-                                DoCloseConnection(info, closeCommander, ConfirmedClosed.Instance);
-                            }
-                            else
-                            {
-                                if (_traceLogging) Log.Debug("Read returned end-of-stream, our side not yet closed");
-                                HandleClose(info, closeCommander, PeerClosed.Instance);
-                            }
-                            break;
-                        case ReadResultType.ReadError:
-                            HandleError(info.Handler, new SocketException((int)read.Error));
-                            break;
-                    }
-                }
-                catch (SocketException cause)
+        
+        private void AwaitRegBehaviour()
+        {
+            Receive<Register>(reg =>
+            {
+                _handler = reg.Handler;
+                if (_traceLogging) Log.Debug("[{0}] registered as connection handler", reg.Handler);
+                Context.WatchWith(_handler, HandlerDied.Instance);
+                Context.Unwatch(_commander);
+                _state = _state with { KeepOpenOnPeerClosed = reg.KeepOpenOnPeerClosed, ReadingSuspended = _pullMode, WritingSuspended = false };
+                // set a default close event - if someone hard-kills us we log an aborted
+                _closeInformation = CloseInformation.Single(_handler, Aborted.Instance);
```

Comment:
Key fix for https://github.com/akkadotnet/akka.net/issues/7634 - we always store an `Abort` + `CloseInformation`, so if the actor gets killed via `Context.Stop` or whatever we will _always_ report a reason now.